### PR TITLE
[SPARK-6007][SQL] Add numRows param in DataFrame.show()

### PR DESCRIPTION
**Pull request by jackylk**
It is useful to let the user decide the number of rows to show in DataFrame.show

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -293,9 +295,10 @@ class DataFrame protected[sql](
    *   1983  03    0.410516        0.442194
    *   1984  04    0.450090        0.483521
    * }}}
+   * @param numRows Number of rows to show
    * @group basic
    */
-  def show(): Unit = println(showString())
+  def show(numRows: Int = 20): Unit = println(showString(numRows))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: this won't work in Java. you need to overload show to provide two shows.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -272,9 +272,9 @@ def isLocal(self):
         """
         return self._jdf.isLocal()
 
-    def show(self):
+    def show(self, numRows=20):
         """
-        Print the first 20 rows.
+        Print the first n rows.
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: there isn't n here. maybe rename numRows -> n

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 3]
```diff
@@ -159,9 +159,11 @@ class DataFrame protected[sql](
 
   /**
    * Internal API for Python
+   * @param numRows Number of rows to show
    */
-  private[sql] def showString(): String = {
-    val data = take(20)
+  private[sql] def showString(numRows: Int = 20): String = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: there is no reason to set a default param value, just have showString(numRows: Int).

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 4]
```diff
@@ -159,9 +159,11 @@ class DataFrame protected[sql](
...
+    val size = count()
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: this forces the engine to read the data twice. why not just run take directly?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 1 above)
Comment (jackylk) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: I added a test case in Java, it seems working. Is there any problem?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: did you try just calling show?

Comment (jackylk) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: I tried 
``` 
     DataFrame df = context.table("testData");
     df.show(10);
     df.show(1000);
     df.select("*").show(30);
```
Is that what you mean?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: can you try df.show()?

in either case, I wouldn't not want to rely on Scala's generated code for default param values for bytecode level binary compatibility. Please implement two show methods, one with an int param and one without. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 5]
```diff
@@ -293,9 +294,15 @@ class DataFrame protected[sql](
...
+   * @param numRows Number of rows to show
    * @group basic
    */
-  def show(): Unit = println(showString())
+  def show(numRows: Int = 20): Unit = println(showString(numRows))
+
+  /**
+   * Displays the [[DataFrame]] in a tabular form. (For Java compatibility)
+   */
+  def show(): Unit = println(showString(20))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: what I mean is... 

```scala
   /**
    * Displays the [[DataFrame]] in a tabular form. For example:
    * {{{
    *   year  month AVG('Adj Close) MAX('Adj Close)
    *   1980  12    0.503218        0.595103
    *   1981  01    0.523289        0.570307
    *   1982  02    0.436504        0.475256
    *   1983  03    0.410516        0.442194
    *   1984  04    0.450090        0.483521
    *   ...
    * }}}
    * @param numRows Number of rows to show
    * @group basic
    */
  def show(numRows: Int): Unit = println(showString(numRows))

  /**
   * Displays the top 20 rows of the [[DataFrame]] in a tabular form.
   */
  def show(): Unit = show(20)
```

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: i.e don't include "for java compatibility" in the user facing doc, and don't include the default param value.

Comment (jackylk) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Thanks a lot

File: python/pyspark/sql/dataframe.py [hunk 6]
```diff
@@ -272,9 +272,9 @@ def isLocal(self):
...
         return self._jdf.isLocal()
 
-    def show(self):
+    def show(self, n=20):
         """
-        Print the first 20 rows.
+        Print the first n rows.
 
         >>> df
         DataFrame[age: int, name: string]
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: one more thing. can you add one more docstring test to test n = 1?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 7]
```diff
@@ -441,4 +441,8 @@ class DataFrameSuite extends QueryTest {
     checkAnswer(df.select(df("key")), testData.select('key).collect().toSeq)
   }
 
+  test("show") {
+    testData.select($"*").show()
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: actually this is a bad idea. it prints a lot of garbage to console. maybe we can ignore the test and add a comment saying this is ignored but we want to make sure it compiles correctly. do the same for java

Comment (jackylk) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Yes, I realized it is not good, it's not checking any answer. I will modify it later

BR,
Jacky

- ·¢×Ô Smartisan T1 -

2015Äê2ÔÂ26ÈÕ£¬ÏÂÎç2:51ÓÚ Reynold Xin <notifications@github.com</<mailto:notifications@github.com>

In sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala<https://github.com/apache/spark/pull/4767#discussion_r25408008>:

> @@ -441,4 +441,8 @@ class DataFrameSuite extends QueryTest {
>      checkAnswer(df.select(df("key")), testData.select('key).collect().toSeq)
>    }
>
> +  test("show") {
> +    testData.select($"*").show()


actually this is a bad idea. it prints a lot of garbage to console. maybe we can ignore the test and add a comment saying this is ignored but we want to make sure it compiles correctly. do the same for java

¡ª

**Comment by rxin** [comment]
Thanks. I've merged this.

**Pull request closed without merging**